### DISCUSSION TITLE
Apache NetBeans, Java version, and Gradle version.

### DISCUSSION THREAD

**Discussion by jmccay-work** [apache/netbeans#6115] [Q&A]
I am currently using both NetBeans versions 12 and 16.  I think Apache NetBeans is too tied to a specific version of Java and Gradle.  I will see if I can explain this properly.

   I understand that Apache NetBeans is built with a specific version of the Java SDK, and that is fine because I think it ships with that version.   Unfortunately, the company I work for still has to use Java 8 for numerous reasons.  Currently, there is no way to change the default version of the Java SDK for all new projects.  Since I can not specify a new default version of Java that is applied to an existing project that I just opened for the first time, I must first manually change the Java version from Java 17 to Java 8.  Then Gradle needs to be compiled with a priming build before the IDE will properly recognize the project and subprojects.  This also has to be done for each project and subproject.
   
   This begs the question of what constitutes opening a given project for the first time?  The term first time here is misleading because it really means the first time with all the current settings.  I recently switched Gradle versions from 6.8.2 to 7.5 by installing Gradle 7.5 onto my system.  I set the Gradle version under _**Tools > Options > Java > Gradle > Gradle Distribution > Custom**_.  

   Just changing versions of Gradle required me to reset the all the Java versions of a project and its subprojects that I had previously opened already in the IDE.  Even if the project was closed when I made the change.
   
   It would be great if a default version of the Java SDK could be set differently than the one shipped with the IDE.  Not all customers can upgrade Java versions when then want to use a new IDE or just a new version of the same IDE.  I think this would decouple the IDE from the Java version it ships with at release time.  The tight coupling between the Java SDK shipped with the IDE and the NetBeans IDE for use with projects was not really a problem with a slower Java release schedule.  Now, there is a new version of Java about every 6 months.

   Thank you.

**Comment by neilcsmith-net** [comment]
Apache NetBeans isn't shipped with a JDK (by default).

The Maven support does have the ability to set a different registered JDK as the default.  The UI is slightly quirky, in that it will show the wrong JDK in the project itself, but this might be something to consider elsewhere.

However, in terms of decoupling, this might better belong in the build systems themselves, specifically toolchain support?  Anything that runs differently inside and outside of NetBeans is perhaps a problem anyway?!

**Reply by jmccay-work** [reply]
Officially, the package I installed on my WSL2 Linux Mint install had a Java 17 SDK included with it.  When I renamed the directory containing the JDK 17, the JDK 17 as default Java platform disappeared from the options list.  Removing all other Java versions seemed to be the way to set the default to Java 8.

Thank you for your help.

**Reply by neilcsmith-net** [reply]
@jmccay-work NetBeans does not support running the IDE on Java 8. That breaks things now, and it will break harder in NB19.

**Comment by emaphis** [comment]
If you absolutely must run on JDK 8, I believe that Netbeans 13 was the latest version to explicitly support running on JDK 8.

**Comment by jmccay-work** [comment]
I figured out how to get Java 8 as the default on Linux Mint.  I renamed the directory with Java 17 in it.  Since it can't find it, it no longer includes it in the libraries, and Java 8 is now the default.  NetBeans still loads.

**Reply by neilcsmith-net** [reply]
@jmccay-work yes, it may still run, but it will not load everything correctly, and has not been supported since NetBeans 13. As mentioned above, it will start to break harder from NetBeans 19.  This is not the way to achieve what you're trying to do!

**Comment by lkishalmi** [comment]
I hope this change in NB20 would help: https://github.com/apache/netbeans/pull/6516
